src/transform/pre_reduce.rs

```diff
@@ -19,39 +76,85 @@ impl run::AsDef for InertDef {
   }
 }
 
-impl Book {
-  /// Reduces the definitions in the book individually, except for the skipped
-  /// ones.
-  ///
-  /// Defs that are not in the book are treated as inert defs.
-  pub fn pre_reduce(&mut self, skip: &dyn Fn(&str) -> bool, max_memory: usize, max_rwts: u64) -> Result<(), String> {
-    let mut host = Host::default();
-    // When a ref is not found in the `Host`, then
-    // put an inert def in its place
-    host.insert_book_with_default(self, &mut |_| DefRef::Owned(Box::new(Def::new(LabSet::ALL, InertDef::default()))));
-    let area = run::Heap::new_bytes(max_memory);
-
-    for (nam, net) in self.nets.iter_mut() {
-      // Skip unnecessary work
-      if net.redexes.is_empty() || skip(nam) {
-        continue;
-      }
+/// State of the pre-reduction algorithm.
+///
+/// Here's how it works:
+/// - Each definition is visited in topological order (dependencies before
+///   dependents). In the case of cycles, one will be arbitrarily selected to be
+///   first.
+/// - The definition is reduced in a [`run::Net`]
+/// - The reduced [`run::Net`] is readback into an [`ast::Net`]
+/// - The [`ast::Net`] is encoded into a [`Vec<Instruction>`]
+/// - The [`ast::Net`] is stored in the [`State`], as it will be used later.
+/// - The [`InterpretedDef`] corresponding to the definition is mutated in-place
+///   and its instructions are replaced with the generated [`Vec<Instruction>`]
+///
+/// At the end, each mutated [`ast::Net`] is placed into the [`Book`],
+/// overriding the previous one.
```

Comment:
this doc comment is useful context for the whole file; can you move it to be part of the module documentation?